### PR TITLE
Fix torch derivative bug

### PR DESCRIPTION
**Pull request by frostedoyster**
Attempts to solve #59

### CODE REVIEW

**Comment by frostedoyster** [comment]
The problem is solved, and the fix is probably more elegant than our previous solution. However, the new construction prevents us from sending any error messages if the user wants second derivatives with respect to `xyz` but didn't declare it at class initialization. Their second derivatives will simply be wrong.

**Comment by Luthaf** [comment]
I guess the issue was that `gradgrad_wrt_grad_out` was not properly computed before, is that right?

File: sphericart-torch/python/tests/test_nn.py [hunk 1]
```diff
@@ -36,5 +36,59 @@ def forward(self, positions):
...
+    xyz_false = xyz.detach().clone()
+    print(nn_false.linear_layer.weight)
+    energy_false, forces_false = nn_false(xyz_false)
+    loss_false = (target - energy_false) ** 2 + torch.sum(
+        (d_target - forces_false) ** 2
+    )
+    loss_false.backward()
+
+    nn_true = NN(backward_second_derivatives=True)
+    print(nn_true.linear_layer.weight)
```
Comment (Luthaf) [inline sphericart-torch/python/tests/test_nn.py:86]:
```suggestion
```

File: sphericart-torch/python/tests/test_nn.py [hunk 2]
```diff
@@ -36,5 +36,59 @@ def forward(self, positions):
...
+                energy,
+                positions,
+                retain_graph=True,
+                create_graph=True,
+            )[0]
+            return energy, forces
+
+    nn_false = NN(backward_second_derivatives=False)
+    xyz_false = xyz.detach().clone()
+    print(nn_false.linear_layer.weight)
```
Comment (Luthaf) [inline sphericart-torch/python/tests/test_nn.py:78]:
```suggestion
```

**Review by Luthaf — APPROVED** [review]
The fix looks good!

File: sphericart-torch/src/autograd.cpp [hunk 3]
```diff
@@ -387,22 +363,32 @@ torch::autograd::variable_list SphericalHarmonicsAutogradBackward::backward(
...
-            grad_2_out.reshape({n_samples, 1, 3})*torch::sum(grad_out.reshape({n_samples, 1, 1, n_sph})*ddsph, 3),
-            2
-        );
-        // the above does the same as the following (but faster):
-        // gradgrad_wrt_xyz = torch::einsum("sa, sk, sabk -> sb", {grad_2_out, grad_out, ddsph});
+        // gradgrad_wrt_xyz is needed to differentiate twice with respect to xyz. However, we only do this if the user
+        // requested it when creating the class
+        if (double_backward) {
+            int n_samples = xyz.sizes()[0];
+            int n_sph = grad_out.sizes()[1];
```
[the remark this answers is not in the document]
Comment (frostedoyster) [reply sphericart-torch/src/autograd.cpp:381]: How confident are you that this works? I've never tried. Should we change it throughout?

[the remark this answers is not in the document]
Comment (Luthaf) [reply sphericart-torch/src/autograd.cpp]: Very, I'm using this in equistore-torch =). And yeah, we could change it where relevant!

**Comment by Luthaf** [comment]
Closing in favor of #63

**Pull request closed without merging**